src/bridge/bridge.c

```diff
@@ -174,145 +174,71 @@ send_init_command (CockpitTransport *transport,
-      goto out;
-    }
-  if (g_unlink (bind_address) < 0)
+  /* get the first line of output to figure out the socket address */
+  g_autoptr (GDataInputStream) stream = g_data_input_stream_new (g_subprocess_get_stdout_pipe (process));
+  g_autofree gchar *first_line = g_data_input_stream_read_line (stream, NULL, NULL, &error);
+  if (!first_line)
     {
-      g_warning ("couldn't remove temporary socket file: %s", g_strerror (errno));
-      goto out;
+      g_warning ("couldn't read address from %s: %s", argv[0], error->message);
+      return NULL;
     }
 
-  if (!g_spawn_sync (NULL, agent_argv, NULL,
-                     G_SPAWN_SEARCH_PATH, setup_ssh_agent,
-                     NULL,
-                     &agent_output, &agent_error,
-                     &status, &error))
-    {
-      if (g_error_matches (error, G_SPAWN_ERROR, G_SPAWN_ERROR_NOENT))
-        g_debug ("couldn't start %s: %s", agent_argv[0], error->message);
-      else
-        g_warning ("couldn't start %s: %s", agent_argv[0], error->message);
-      goto out;
-    }
+  g_autoptr(GRegex) regex = g_regex_new (socket_pattern, G_REGEX_RAW, 0, &error);
+  g_assert_no_error (error);
 
-  if (!g_spawn_check_exit_status (status, &error))
+  g_autoptr(GMatchInfo) info = NULL;
+  if (!g_regex_match (regex, first_line, 0, &info))
     {
-      g_warning ("couldn't start %s: %s: %s", agent_argv[0],
-                 error->message, agent_error);
-      goto out;
+      g_warning ("output from %s didn't match expected pattern %s", argv[0], socket_pattern);
+      return NULL;
     }
 
-  pid_line = strstr (agent_output, "SSH_AGENT_PID=");
-  if (pid_line)
-    {
-      if (sscanf (pid_line, "SSH_AGENT_PID=%d;", &pid) != 1)
-        {
-            g_warning ("couldn't find pid in %s", pid_line);
-            goto out;
-        }
-    }
+  g_autofree gchar *socket_address = g_match_info_fetch (info, 1);
+  cockpit_setenv_check (socket_envvar, socket_address, TRUE);
 
-  if (pid < 1)
-    {
-      g_warning ("couldn't get agent pid from ssh-agent output: %s", agent_output);
-      goto out;
-    }
+  return g_steal_pointer (&process);
+}
 
-  g_debug ("launched %s", agent_argv[0]);
-  cockpit_setenv_check ("SSH_AUTH_SOCK", bind_address, TRUE);
-
-out:
-  g_clear_error (&error);
-  if (fd >= 0)
-    close (fd);
-  g_free (bind_address);
-  g_free (agent_error);
-  g_free (agent_output);
-  return pid;
+static GSubprocess *
+start_dbus_daemon (void)
+{
+  const char * const cmd[] = { "dbus-daemon", "--print-address", "--session", NULL };
+  return start_helper_process (cmd, "^(.*$)", "DBUS_SESSION_BUS_ADDRESS");
 }
 
-static gboolean
-have_env (const gchar *name)
+static GSubprocess *
+start_ssh_agent (void)
 {
-  const gchar *env = g_getenv (name);
-  return env && env[0];
+  const char * const cmd[] = { "ssh-agent", "-s", "-D", NULL };
+  return start_helper_process (cmd, "SSH_AUTH_SOCK=([^;]*);", "SSH_AUTH_SOCK");
```

Comment:
I'm quite sure you need to escape the () here as well, just like you did in start_dbus_daemon().

Reply:
i don't use `\` anywhere, actually.